salt/modules/heat.py

```diff
@@ -0,0 +1,870 @@
+            or 'AWSTemplateFormatVersion' in tpl):
+        raise ValueError(("Template format version not found."))
+    return tpl
+
+
+def _parse_enviroment(env_str):
+    '''
+    Parsing template
+    '''
+    try:
+        env = yaml.load(env_str, Loader=YamlLoader)
+    except yaml.YAMLError:
+        try:
+            env = yaml.load(env_str, Loader=yaml.SafeLoader)
+        except yaml.YAMLError as yea:
+            raise ValueError(yea)
+    else:
+        if env is None:
+            env = {}
+        elif not isinstance(env, dict):
+            raise ValueError(('The environment is not a valid '
+                              'YAML mapping data type.'))
+
+    for param in env:
+        if param not in SECTIONS:
+            raise ValueError(('environment has wrong section "{0}"').format(param))
+
+    return env
+
+
+def _get_stack_events(h_client, stack_id, event_args):
+    '''
+    Get event for stack
+    '''
+    event_args['stack_id'] = stack_id
+    event_args['resource_name'] = None
+    try:
+        events = h_client.events.list(**event_args)
+    except exc.HTTPNotFound as ex:
+        raise exc.CommandError(str(ex))
+    else:
+        for event in events:
+            event.stack_name = stack_id.split("/")[0]
+        return events
+
+
+def _poll_for_events(h_client, stack_name, action=None, poll_period=5, marker=None):
+    '''
+    Polling stack events
+    '''
+    if action:
+        stop_status = ('{0}_FAILED'.format(action), '{0}_COMPLETE'.format(action))
+        stop_check = lambda a: a in stop_status
+    else:
+        stop_check = lambda a: a.endswith('_COMPLETE') or a.endswith('_FAILED')
+
+    no_event_polls = 0
+    msg_template = ("\n Stack %(name)s %(status)s \n")
+    while True:
+        events = _get_stack_events(h_client, stack_id=stack_name,
+                                   event_args={'sort_dir': 'asc', 'marker': marker})
+
+        if len(events) == 0:
+            no_event_polls += 1
+        else:
+            no_event_polls = 0
+            # set marker to last event that was received.
+            marker = getattr(events[-1], 'id', None)
+            for event in events:
+                # check if stack event was also received
+                if getattr(event, 'resource_name', '') == stack_name:
+                    stack_status = getattr(event, 'resource_status', '')
+                    msg = msg_template % dict(
+                        name=stack_name, status=stack_status)
+                    if stop_check(stack_status):
+                        return stack_status, msg
+
+        if no_event_polls >= 2:
+            # after 2 polls with no events, fall back to a stack get
+            stack = h_client.stacks.get(stack_name)
+            stack_status = stack.stack_status
+            msg = msg_template % dict(
+                name=stack_name, status=stack_status)
+            if stop_check(stack_status):
+                return stack_status, msg
```

Comment:
I can't speak to the effects of an infinite loop there but in Salt, an infinite loop like this will basically hang a minion job and prevent it from returning results. What do you think about at least adding a counter so that it fails eventually?